File: packages/core-go/src/room.go
package main

import "github.com/google/uuid"

type User struct {
	ID   string `json:"id"`
	Name string `json:"name"`
}

type Answer struct {
	ID     string `json:"id"`
	Text   string `json:"text"`
	Player *User  `json:"player"`
}

type Question struct {
	ID              string `json:"id"`
	QuestionnaireID string `json:"questionnaireId"`
	Text            string `json:"text"`
}

type Round struct {
	Number   uint16    `json:"number"`
	Leader   *User     `json:"leader"`
	Question *Question `json:"question"`
	Answers  []*Answer `json:"answer"`
}

type Room struct {
	ID              string  `json:"id"`
	Host            *User   `json:"host"`
	Code            string  `json:"code"`
	Players         []*User `json:"players"`
	CurrentStage    string  `json:"currentStage"`
	Round           *Round  `json:"round"`
	QuestionnaireId string  `json:"questionnaireId"`
	RoundPerPlayer  int     `json:"roundPerPlayer"`
	RoundDuration   int     `json:"roundDuration"`
}

type CreateRoomProps struct {
	Username        string
	RoundsPerPlayer int
	RoundDuration   int
	Code            string
}

func CreateRoom(props *CreateRoomProps) *Room {
	room := &Room{
		ID: uuid.NewString(),
		Host: &User{
			ID:   uuid.NewString(),
			Name: props.Username,
		},
		Code:           props.Code,
		CurrentStage:   "",
		RoundPerPlayer: props.RoundsPerPlayer,
		RoundDuration:  props.RoundDuration,
	}

	return room
}

File: packages/ui/src/models/room.ts
type ID = string;

export interface I_User {
    id: ID;
    name: string;
}

enum StageEnum {
    wait = "wait",
    prepare = "prepare",
    play = "play",
    vote = "vote",
    result = "result"
}

enum StatusEnum {
    play = "play",
    finished = "finished"
}

export interface I_Answer {
    question: I_Question;
    option: I_Option;
    player: I_User;
}

export interface I_Round {
    number: number;
    leader: I_User;
    winner: I_Answer;
    question: I_Question;
    answers: I_Answer[];
    current_stage: StageEnum;
}

export interface I_Question {
    id: ID;
    text: string;
}

export interface I_Option {
    id: ID;
    text: string;
}

export interface I_Questionnaire {
    id: ID;
    name: string;
    author: ID;
    questions: I_Question[];
    options: I_Option[];
}

export interface I_Leaderboard {
    [userId: string]: number;
}

export interface I_RoomInfo {
    id: string;
    host: I_User;
    round_duration: number;
    rounds_per_player: number;
    players: I_User[];
    round: I_Round;
    questions: I_Question[];
    leaderboard: I_Leaderboard;
    code: number;
    status: StatusEnum;
}